### PR TITLE
Allow list search in CONTEXT

### PR DESCRIPTION
**Pull request by lmcalvo**
Captura de pantalla 2023-06-15 a las 12 09 51

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -6,6 +6,8 @@ v3.1.0
 
 *Release date: In development*
 
+- Update `[CONTEXT:a.b.c]` to allow list access in context, f.e: context.list.1 access to the second element of list
```
Comment (pabloge) [inline CHANGELOG.rst:9]:
```suggestion
- Update `[CONTEXT:a.b.c]` replacement to allow lists access, e.g. [CONTEXT:list.1] to access the second element of list
```

File: toolium/test/utils/test_dataset_map_param_context.py [hunk 2]
```diff
@@ -277,3 +277,172 @@ class Context(object):
     with pytest.raises(Exception) as excinfo:
         map_param("[CONTEXT:one.three]")
     assert f"'three' key not found in {context.one} value in context" == str(excinfo.value)
+
+
+def test_a_context_param_list():
+    """
+    Verification of a list without index as CONTEXT to get all the list, not a element in the list
```
Comment (pabloge) [inline toolium/test/utils/test_dataset_map_param_context.py:284]:
```suggestion
    Verification of a list without index as CONTEXT to get all the list, not an element in the list
```

File: toolium/test/utils/test_dataset_map_param_context.py [hunk 3]
```diff
@@ -277,3 +277,172 @@ class Context(object):
...
+    assert "Invalid index '2', list size is '2'. 2 >= 2." == str(excinfo.value)
+
+    with pytest.raises(Exception) as excinfo:
+        map_param("[CONTEXT:list.cmsScrollableActions.5.id]")
+    assert "Invalid index '5', list size is '2'. 5 >= 2." == str(excinfo.value)
+
+
+def test_a_context_param_list_no_numeric_index():
+    """
+    Verification of a list with a no numeric index as CONTEXT
```
Comment (pabloge) [inline toolium/test/utils/test_dataset_map_param_context.py:392]:
```suggestion
    Verification of a list with a non-numeric index as CONTEXT
```

File: toolium/test/utils/test_dataset_map_param_context.py [hunk 4]
```diff
@@ -277,3 +277,172 @@ class Context(object):
...
+    assert "the index 'prueba' must be a numeric index" == str(excinfo.value)
+
+    with pytest.raises(Exception) as excinfo:
+        map_param("[CONTEXT:list.cmsScrollableActions.'36'.id]")
+    assert "the index ''36'' must be a numeric index" == str(excinfo.value)
+
+
+def test_a_context_param_class_no_numeric_index():
+    """
+    Verification of a class with a no numeric index as CONTEXT
```
Comment (pabloge) [inline toolium/test/utils/test_dataset_map_param_context.py:423]:
```suggestion
    Verification of a class with a non-numeric index as CONTEXT
```

File: toolium/utils/dataset.py [hunk 5]
```diff
@@ -581,6 +581,8 @@ def get_value_from_context(param, context):
     storages or the context object itself. In a dotted case, "last_request.result" is searched as a "last_request" key
     in the context storages or as a property of the context object whose name is last_request. In both cases, when
     found, "result" is considered (and resolved) as a property or a key into the returned value.
+    If the element is a list then return the first element as default or take the index given in the next part
+    f.e: list.1 returns the second element of the list 'list'.
```
Comment (pabloge) [inline toolium/utils/dataset.py:585]:
```suggestion
    If the resolved element at one of the tokens is a list, then the next token (if present) is used as the index
    to select one of its elements, e.g. "list.1" returns the second element of the list "list".
```

File: toolium/utils/dataset.py [hunk 6]
```diff
@@ -594,22 +596,42 @@ def get_value_from_context(param, context):
...
+            value = value[int(part)]
         elif hasattr(value, part):
             value = getattr(value, part)
         else:
-            if isinstance(value, dict):
-                msg = f"'{part}' key not found in {value} value in context"
-            else:
-                msg = f"'{part}' attribute not found in {type(value).__name__} class in context"
+            msg = get_value_context_errors(value, part)
+        if msg:
```
Comment (pabloge) [inline toolium/utils/dataset.py:610]: This `if` is no longer needed: `msg` is always set in the previous line (and only there).
```suggestion
```

File: toolium/utils/dataset.py [hunk 7]
```diff
@@ -594,22 +596,42 @@ def get_value_from_context(param, context):
...
     return value
 
 
+def get_value_context_errors(value, part):
+    """
+    Returns the errors
+
+    :param value: value of the CONTEXT dict
+    :param part: part to take in the value
+    :return: errors
```
Comment (pabloge) [inline toolium/utils/dataset.py:622]:
```suggestion
    Returns an appropriate error message when an error occurs resolving a CONTEXT reference.

    :param value: last value that has been correctly resolved
    :param part: token that is causing the error
    :return: a string with the error message
```

File: toolium/utils/dataset.py [hunk 8]
```diff
@@ -594,22 +596,42 @@ def get_value_from_context(param, context):
...
+        elif isinstance(value, list) and part.isdigit() and int(part) < len(value):
+            value = value[int(part)]
         elif hasattr(value, part):
             value = getattr(value, part)
         else:
-            if isinstance(value, dict):
-                msg = f"'{part}' key not found in {value} value in context"
-            else:
-                msg = f"'{part}' attribute not found in {type(value).__name__} class in context"
+            msg = get_value_context_errors(value, part)
```
Comment (pabloge) [inline toolium/utils/dataset.py:609]: I would rather make it private, it does not seem to be very reusable outside this module.
```suggestion
            msg = _get_value_context_error_msg(value, part)
```

File: toolium/utils/dataset.py [hunk 9]
```diff
@@ -594,22 +596,42 @@ def get_value_from_context(param, context):
...
+        if msg:
             logger.error(msg)
             raise Exception(msg)
     return value
 
 
+def get_value_context_errors(value, part):
```
Comment (pabloge) [inline toolium/utils/dataset.py:616]: I would rather make it private, it does not seem very reusable outside this module.
```suggestion
def _get_value_context_error_msg(value, part):
```

**Review by pabloge — CHANGES_REQUESTED** [review]

**Review by pabloge — APPROVED** [review]
LGTM

**Review by rgonalo — APPROVED** [review]

**Pull request merged**